File: articles/design-thinking-organizational-innovation/c/innovation_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ITEMS 160
#define MAX_LINE 640
#define MAX_NAME 220

typedef struct {
    char name[MAX_NAME];
    double desirability;
    double feasibility;
    double viability;
    double equity;
    double learning_value;
    double implementation_readiness;
    double risk;
    double design_value;
} InnovationConcept;

double compute_design_value(const InnovationConcept *x) {
    return 0.22 * x->desirability +
           0.16 * x->feasibility +
           0.16 * x->viability +
           0.18 * x->equity +
           0.12 * x->learning_value +
           0.10 * x->implementation_readiness -
           0.06 * x->risk;
}

int compare_desc(const void *a, const void *b) {
    const InnovationConcept *ia = (const InnovationConcept *)a;
    const InnovationConcept *ib = (const InnovationConcept *)b;

    if (ia->design_value < ib->design_value) return 1;
    if (ia->design_value > ib->design_value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/organizational_innovation_concepts_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    InnovationConcept concepts[MAX_ITEMS];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_ITEMS) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(concepts[count].name, token, MAX_NAME - 1);
        concepts[count].name[MAX_NAME - 1] = '\0';

        strtok(NULL, ","); /* concept_type */
        strtok(NULL, ","); /* organizational_domain */

        token = strtok(NULL, ",");
        concepts[count].desirability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].feasibility = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].viability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].equity = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].learning_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].implementation_readiness = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        concepts[count].risk = token ? atof(token) : 0.0;

        concepts[count].design_value = compute_design_value(&concepts[count]);

        count++;
    }

    fclose(file);

    qsort(concepts, count, sizeof(InnovationConcept), compare_desc);

    printf("rank,concept,design_value,risk\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f,%.6f\n", i + 1, concepts[i].name, concepts[i].design_value, concepts[i].risk);
    }

    return 0;
}

File: articles/design-thinking-organizational-innovation/go/innovation_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type InnovationConcept struct {
	Name                    string
	Desirability            float64
	Feasibility             float64
	Viability               float64
	Equity                  float64
	LearningValue           float64
	ImplementationReadiness float64
	Risk                    float64
	DesignValue             float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func designValue(x InnovationConcept) float64 {
	return 0.22*x.Desirability +
		0.16*x.Feasibility +
		0.16*x.Viability +
		0.18*x.Equity +
		0.12*x.LearningValue +
		0.10*x.ImplementationReadiness -
		0.06*x.Risk
}

func loadConcepts(path string) []InnovationConcept {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []InnovationConcept

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 10 {
			log.Fatalf("malformed row: %v", row)
		}

		x := InnovationConcept{
			Name:                    row[0],
			Desirability:            parseFloat(row[3]),
			Feasibility:             parseFloat(row[4]),
			Viability:               parseFloat(row[5]),
			Equity:                  parseFloat(row[6]),
			LearningValue:           parseFloat(row[7]),
			ImplementationReadiness: parseFloat(row[8]),
			Risk:                    parseFloat(row[9]),
		}
		x.DesignValue = designValue(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/organizational_innovation_concepts_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadConcepts(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].DesignValue > items[j].DesignValue
	})

	fmt.Println("rank,concept,design_value,risk")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.DesignValue, x.Risk)
	}
}
